pkgs/stdenv/darwin/override-sdk.nix

```diff
@@ -0,0 +1,412 @@
+      let
+        replacement = replacePropagatedPackages newPackages pkg;
+        outPath = builtins.unsafeDiscardStringContext pkg;
+      in
+      if pkg == null || newPackages ? ${outPath} then
+        newPackages
+      else
+        newPackages // { ${outPath} = replacement; }
+    ) baseMapping input;
+
+  overrideSDK =
+    stdenv: sdkVersion:
+    let
+      newVersion = {
+        inherit (stdenv.hostPlatform) darwinMinVersion darwinSdkVersion;
+      } // (if lib.isAttrs sdkVersion then sdkVersion else { darwinSdkVersion = sdkVersion; });
+
+      inherit (newVersion) darwinMinVersion darwinSdkVersion;
+
+      # Used to get an SDK version corresponding to the requested `darwinSdkVersion`.
+      # TODO: Treat `darwinSdkVersion` as a constraint rather than as an exact version.
+      resolveSDK = pkgs: pkgs.darwin."apple_sdk_${lib.replaceStrings [ "." ] [ "_" ] darwinSdkVersion}";
+
+      # `newSdkPackages` is constructed based on the assumption that SDK packages only
+      # propagate versioned packages from that SDK -- that they neither propagate
+      # unversioned packages SDK packages nor propagate non-SDK packages (such as curl).
+      #
+      # Note: `builtins.unsafeDiscardStringContext` is used to allow the path from the
+      # original package output to be mapped to the replacement. This is safe because
+      # the value is not persisted anywhere and necessary because store paths are not
+      # allowed as attrset names otherwise.
+      baseSdkMapping =
+        let
+          sdkPackages = lib.genericClosure {
+            startSet = lib.pipe args [
+              (lib.flip removeAttrs [
+                "lib"
+                "extendMkDerivationArgs"
+              ])
+              (lib.filterAttrs (_: lib.hasAttr "darwin"))
+              (lib.mapAttrs (
+                name: value: {
+                  key = name;
+                  inherit value;
+                }
+              ))
+              lib.attrValues
+            ];
+            operator =
+              item:
+              let
+                newSDK = resolveSDK item.value;
+              in
+              if item ? replacement then
+                [ ]
+              else if item ? replacementList then
+                item.replacementList
+              else
+                [
+                  {
+                    key = "frameworks";
+                    replacementList = mkMapping "frameworks" item.value newSDK;
+                  }
+                  {
+                    key = "libs";
+                    pkgs = mkMapping "libs" item.value newSDK;
+                  }
+                  {
+                    key = "overrides";
+                    replacementList = mkOverrides item.value newSDK newVersion;
+                  }
+                ];
+          };
+        in
+        lib.pipe sdkPackages [
+          (lib.filter (attrs: attrs ? replacement))
+          (map (
+            { key, replacement }:
```

Comment:
The sets here only have two keys: `key` and one of `pkgs, `replacement`, or `replacementList`. It’s not an issue because the filter should remove those other two. If one of those creeps in, it’s an error, and evaluation should fail. The only reason for the remapping is make everything conform to `listToAttrs`.

One possibility is defining `name` as an alias for `key` and rename `replacement` to `value`. That should let me drop the remapping step. Would that be better?

Reply:
No, the current way is fine. I just wasn't certain the attributes were exclusive.